Optimizations for invalidate tags

arichardson opened the pull request:
RISCV boot benchmark before:
```
Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.pre-invalidate -kernel /local/scratch/alr48/cheri/output/QEMU_BENCHMARK_RISCV64_HYBRID_KERNEL_PURECAP_MFS_ROOT -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -device virtio-net-device,netdev=net0 -netdev user,id=net0,ipv6=off -append init_path=/sbin/startup-benchmark.sh' (10 runs):

       4190.311225      task-clock (msec)         #    0.885 CPUs utilized            ( +-  0.65% )
             1,230      context-switches          #    0.294 K/sec                    ( +-  0.59% )
                 2      cpu-migrations            #    0.001 K/sec                    ( +- 34.93% )
             7,450      page-faults               #    0.002 M/sec                    ( +-  0.59% )
    13,984,847,818      cycles                    #    3.337 GHz                      ( +-  0.64% )  (30.71%)
    40,573,002,845      instructions              #    2.90  insn per cycle           ( +-  0.31% )  (38.52%)
     5,790,907,624      branches                  # 1381.976 M/sec                    ( +-  0.19% )  (38.49%)
        57,394,474      branch-misses             #    0.99% of all branches          ( +-  0.52% )  (38.58%)
    10,027,283,827      L1-dcache-loads           # 2392.969 M/sec                    ( +-  0.32% )  (38.70%)
       119,255,313      L1-dcache-load-misses     #    1.19% of all L1-dcache hits    ( +-  0.93% )  (38.62%)
         3,900,859      LLC-loads                 #    0.931 M/sec                    ( +-  2.04% )  (30.81%)
           706,916      LLC-load-misses           #   18.12% of all LL-cache hits     ( +-  9.14% )  (30.76%)
   <not supported>      L1-icache-loads
       342,126,170      L1-icache-load-misses                                         ( +-  0.56% )  (30.68%)
     9,987,088,685      dTLB-loads                # 2383.376 M/sec                    ( +-  0.83% )  (30.67%)
           209,036      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +-  3.75% )  (30.70%)
         3,621,255      iTLB-loads                #    0.864 M/sec                    ( +-  1.31% )  (30.66%)
         1,077,941      iTLB-load-misses          #   29.77% of all iTLB cache hits   ( +- 11.64% )  (30.62%)

       4.732444569 seconds time elapsed
```
After:
```
Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri -kernel /local/scratch/alr48/cheri/output/QEMU_BENCHMARK_RISCV64_HYBRID_KERNEL_PURECAP_MFS_ROOT -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -device virtio-net-device,netdev=net0 -netdev user,id=net0,ipv6=off -append init_path=/sbin/startup-benchmark.sh' (10 runs):

After:
       3965.398176      task-clock (msec)         #    0.872 CPUs utilized            ( +-  0.41% )
             1,197      context-switches          #    0.302 K/sec                    ( +-  0.51% )
                 3      cpu-migrations            #    0.001 K/sec                    ( +- 20.83% )
             9,886      page-faults               #    0.002 M/sec                    ( +- 22.31% )
    12,975,511,287      cycles                    #    3.272 GHz                      ( +-  0.33% )  (30.96%)
    36,310,805,061      instructions              #    2.80  insn per cycle           ( +-  0.32% )  (38.68%)
     4,290,656,596      branches                  # 1082.024 M/sec                    ( +-  0.31% )  (38.60%)
        52,444,394      branch-misses             #    1.22% of all branches          ( +-  0.41% )  (38.54%)
     9,520,541,071      L1-dcache-loads           # 2400.904 M/sec                    ( +-  0.27% )  (38.44%)
       121,941,572      L1-dcache-load-misses     #    1.28% of all L1-dcache hits    ( +-  0.84% )  (38.36%)
         3,899,538      LLC-loads                 #    0.983 M/sec                    ( +-  6.62% )  (30.56%)
           904,366      LLC-load-misses           #   23.19% of all LL-cache hits     ( +- 21.52% )  (30.59%)
   <not supported>      L1-icache-loads
       338,198,675      L1-icache-load-misses                                         ( +-  0.74% )  (30.66%)
     9,544,742,562      dTLB-loads                # 2407.007 M/sec                    ( +-  0.30% )  (30.69%)
           220,078      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +- 11.12% )  (30.84%)
         4,496,137      iTLB-loads                #    1.134 M/sec                    ( +-  1.62% )  (30.87%)
           681,771      iTLB-load-misses          #   15.16% of all iTLB cache hits   ( +- 15.80% )  (30.90%)

       4.547806136 seconds time elapsed                                          ( +-  0.24% )
```

arichardson committed:
Cache the tag memory buffer in the QEMU iotlb

This avoids a call to qemu_ram_block_from_host() in the commonly used
cheri_tag_invalidate() function. This reduces the number of branches and
cycles quite a lot in a CheriBSD kernel boot benchmark.
This also adds a fake tagmem buffer ALL_ZERO_TAGBLK for tag blocks that
have not been allocated yet and are therefore all zero.

The optimization is currently only used in the invalidate path since that
is the most critical one, the get tag+set tag code still does a full
page table walk.

RISCV boot benchmark before:

Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.pre-invalidate -kernel /local/scratch/alr48/cheri/output/QEMU_BENCHMARK_RISCV64_HYBRID_KERNEL_PURECAP_MFS_ROOT -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -device virtio-net-device,netdev=net0 -netdev user,id=net0,ipv6=off -append init_path=/sbin/startup-benchmark.sh' (10 runs):

       4190.311225      task-clock (msec)         #    0.885 CPUs utilized            ( +-  0.65% )
             1,230      context-switches          #    0.294 K/sec                    ( +-  0.59% )
                 2      cpu-migrations            #    0.001 K/sec                    ( +- 34.93% )
             7,450      page-faults               #    0.002 M/sec                    ( +-  0.59% )
    13,984,847,818      cycles                    #    3.337 GHz                      ( +-  0.64% )  (30.71%)
    40,573,002,845      instructions              #    2.90  insn per cycle           ( +-  0.31% )  (38.52%)
     5,790,907,624      branches                  # 1381.976 M/sec                    ( +-  0.19% )  (38.49%)
        57,394,474      branch-misses             #    0.99% of all branches          ( +-  0.52% )  (38.58%)
    10,027,283,827      L1-dcache-loads           # 2392.969 M/sec                    ( +-  0.32% )  (38.70%)
       119,255,313      L1-dcache-load-misses     #    1.19% of all L1-dcache hits    ( +-  0.93% )  (38.62%)
         3,900,859      LLC-loads                 #    0.931 M/sec                    ( +-  2.04% )  (30.81%)
           706,916      LLC-load-misses           #   18.12% of all LL-cache hits     ( +-  9.14% )  (30.76%)
   <not supported>      L1-icache-loads
       342,126,170      L1-icache-load-misses                                         ( +-  0.56% )  (30.68%)
     9,987,088,685      dTLB-loads                # 2383.376 M/sec                    ( +-  0.83% )  (30.67%)
           209,036      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +-  3.75% )  (30.70%)
         3,621,255      iTLB-loads                #    0.864 M/sec                    ( +-  1.31% )  (30.66%)
         1,077,941      iTLB-load-misses          #   29.77% of all iTLB cache hits   ( +- 11.64% )  (30.62%)

       4.732444569 seconds time elapsed

After:

Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri -kernel /local/scratch/alr48/cheri/output/QEMU_BENCHMARK_RISCV64_HYBRID_KERNEL_PURECAP_MFS_ROOT -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -device virtio-net-device,netdev=net0 -netdev user,id=net0,ipv6=off -append init_path=/sbin/startup-benchmark.sh' (10 runs):

       3774.365088      task-clock (msec)         #    0.858 CPUs utilized            ( +-  0.48% )
             1,142      context-switches          #    0.303 K/sec                    ( +-  0.22% )
                 3      cpu-migrations            #    0.001 K/sec                    ( +- 28.21% )
             7,688      page-faults               #    0.002 M/sec                    ( +-  0.54% )
    12,543,787,636      cycles                    #    3.323 GHz                      ( +-  0.67% )  (30.78%)
    35,564,664,145      instructions              #    2.84  insn per cycle           ( +-  0.16% )  (38.52%)
     4,455,520,915      branches                  # 1180.469 M/sec                    ( +-  0.27% )  (38.50%)
        52,249,006      branch-misses             #    1.17% of all branches          ( +-  0.44% )  (38.53%)
     8,975,263,824      L1-dcache-loads           # 2377.953 M/sec                    ( +-  0.19% )  (38.45%)
       117,688,401      L1-dcache-load-misses     #    1.31% of all L1-dcache hits    ( +-  1.09% )  (38.44%)
         3,960,296      LLC-loads                 #    1.049 M/sec                    ( +-  0.93% )  (30.68%)
           775,589      LLC-load-misses           #   19.58% of all LL-cache hits     ( +-  6.84% )  (30.69%)
   <not supported>      L1-icache-loads
       337,501,992      L1-icache-load-misses                                         ( +-  0.36% )  (30.75%)
     8,854,192,783      dTLB-loads                # 2345.876 M/sec                    ( +-  0.64% )  (30.78%)
           209,355      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +-  2.43% )  (30.81%)
         2,015,672      iTLB-loads                #    0.534 M/sec                    ( +-  2.68% )  (30.77%)
         1,222,479      iTLB-load-misses          #   60.65% of all iTLB cache hits   ( +-  5.90% )  (30.80%)

       4.397370551 seconds time elapsed                                          ( +-  0.40% )

MIPS Before:
Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-cheri128.pre-invalidate -kernel /local/scratch/alr48/cheri/output/QEMU_BENCHMARK_MIPS_HYBRID_KERNEL_PURECAP_MFS_ROOT -M malta -m 2048 -nographic -nic user,id=net0,ipv6=off -append init_path=/sbin/startup-benchmark.sh' (10 runs):

       2121.350785      task-clock (msec)         #    0.778 CPUs utilized            ( +-  0.76% )
               995      context-switches          #    0.469 K/sec                    ( +-  1.46% )
                 2      cpu-migrations            #    0.001 K/sec                    ( +- 25.00% )
            10,844      page-faults               #    0.005 M/sec                    ( +-  0.31% )
     7,009,666,916      cycles                    #    3.304 GHz                      ( +-  0.63% )  (30.55%)
    18,406,270,097      instructions              #    2.63  insn per cycle           ( +-  0.45% )  (38.26%)
     2,740,102,535      branches                  # 1291.678 M/sec                    ( +-  0.54% )  (38.19%)
        27,367,400      branch-misses             #    1.00% of all branches          ( +-  0.67% )  (38.17%)
     4,644,550,334      L1-dcache-loads           # 2189.431 M/sec                    ( +-  0.16% )  (38.24%)
        68,887,607      L1-dcache-load-misses     #    1.48% of all L1-dcache hits    ( +-  0.91% )  (38.38%)
         3,325,827      LLC-loads                 #    1.568 M/sec                    ( +-  2.80% )  (30.65%)
           872,121      LLC-load-misses           #   26.22% of all LL-cache hits     ( +-  9.66% )  (30.71%)
   <not supported>      L1-icache-loads
       190,914,762      L1-icache-load-misses                                         ( +-  0.45% )  (30.81%)
     4,563,856,812      dTLB-loads                # 2151.392 M/sec                    ( +-  0.53% )  (31.07%)
           283,334      dTLB-load-misses          #    0.01% of all dTLB cache hits   ( +-  3.81% )  (31.16%)
         1,200,039      iTLB-loads                #    0.566 M/sec                    ( +-  2.98% )  (31.11%)
           742,040      iTLB-load-misses          #   61.83% of all iTLB cache hits   ( +-  4.64% )  (30.96%)

       2.728018831 seconds time elapsed                                          ( +-  5.86% )

MIPS after:
Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-cheri128 -kernel /local/scratch/alr48/cheri/output/QEMU_BENCHMARK_MIPS_HYBRID_KERNEL_PURECAP_MFS_ROOT -M malta -m 2048 -nographic -nic user,id=net0,ipv6=off -append init_path=/sbin/startup-benchmark.sh' (10 runs):

       1988.293672      task-clock (msec)         #    0.776 CPUs utilized            ( +-  0.68% )
               932      context-switches          #    0.468 K/sec                    ( +-  2.07% )
                 2      cpu-migrations            #    0.001 K/sec                    ( +- 31.68% )
            10,790      page-faults               #    0.005 M/sec                    ( +-  0.04% )
     6,627,621,203      cycles                    #    3.333 GHz                      ( +-  1.01% )  (30.61%)
    16,997,048,979      instructions              #    2.56  insn per cycle           ( +-  0.63% )  (38.48%)
     2,230,206,104      branches                  # 1121.668 M/sec                    ( +-  0.65% )  (38.52%)
        23,617,540      branch-misses             #    1.06% of all branches          ( +-  1.10% )  (38.64%)
     4,314,732,430      L1-dcache-loads           # 2170.068 M/sec                    ( +-  0.39% )  (38.76%)
        69,409,878      L1-dcache-load-misses     #    1.61% of all L1-dcache hits    ( +-  1.32% )  (38.61%)
         3,168,414      LLC-loads                 #    1.594 M/sec                    ( +-  2.25% )  (30.65%)
           745,336      LLC-load-misses           #   23.52% of all LL-cache hits     ( +-  4.91% )  (30.65%)
   <not supported>      L1-icache-loads
       186,915,139      L1-icache-load-misses                                         ( +-  1.00% )  (30.69%)
     4,316,102,316      dTLB-loads                # 2170.757 M/sec                    ( +-  0.51% )  (30.78%)
           273,214      dTLB-load-misses          #    0.01% of all dTLB cache hits   ( +-  3.26% )  (30.83%)
           774,153      iTLB-loads                #    0.389 M/sec                    ( +-  4.79% )  (30.72%)
           912,982      iTLB-load-misses          #  117.93% of all iTLB cache hits   ( +- 13.29% )  (30.55%)

       2.563825348 seconds time elapsed                                          ( +-  6.59% )

arichardson committed:
Handle TLB_WATCHPOINT correctly for MMU_DATA_CAP_STORE

This was previously being treated as a read instead of a write.

arichardson committed:
Avoid a few more branches in cheri_invalidate_tags()

RISCV64 minimal startup benchmark before:
Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.first-invalidate-opt -kernel /local/scratch/alr48/cheri/output/QEMU_BENCHMARK_RISCV64_HYBRID_KERNEL_PURECAP_MFS_ROOT -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -device virtio-net-device,netdev=net0 -netdev user,id=net0,ipv6=off -append init_path=/sbin/startup-benchmark.sh' (10 runs):

       3818.656355      task-clock (msec)         #    0.858 CPUs utilized            ( +-  0.23% )
             1,144      context-switches          #    0.300 K/sec                    ( +-  0.48% )
                 1      cpu-migrations            #    0.000 K/sec                    ( +- 50.92% )
             7,638      page-faults               #    0.002 M/sec                    ( +-  0.46% )
    12,229,798,729      cycles                    #    3.203 GHz                      ( +-  1.57% )  (30.75%)
    35,559,454,379      instructions              #    2.91  insn per cycle           ( +-  0.25% )  (38.44%)
     4,476,918,170      branches                  # 1172.380 M/sec                    ( +-  0.26% )  (38.34%)
        52,007,183      branch-misses             #    1.16% of all branches          ( +-  0.24% )  (38.31%)
     9,010,721,539      L1-dcache-loads           # 2359.658 M/sec                    ( +-  0.12% )  (38.38%)
       114,842,912      L1-dcache-load-misses     #    1.27% of all L1-dcache hits    ( +-  1.35% )  (38.36%)
         3,706,186      LLC-loads                 #    0.971 M/sec                    ( +-  2.18% )  (30.73%)
           506,797      LLC-load-misses           #   13.67% of all LL-cache hits     ( +-  5.17% )  (30.75%)
   <not supported>      L1-icache-loads
       338,123,594      L1-icache-load-misses                                         ( +-  0.44% )  (30.84%)
     8,651,543,246      dTLB-loads                # 2265.599 M/sec                    ( +-  1.76% )  (30.90%)
           201,367      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +-  5.11% )  (30.93%)
         1,970,944      iTLB-loads                #    0.516 M/sec                    ( +-  3.18% )  (30.94%)
         1,300,389      iTLB-load-misses          #   65.98% of all iTLB cache hits   ( +-  6.37% )  (30.78%)

       4.451461056 seconds time elapsed                                          ( +-  0.13% )                                  ( +-  6.59% )

After:
Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri -kernel /local/scratch/alr48/cheri/output/QEMU_BENCHMARK_RISCV64_HYBRID_KERNEL_PURECAP_MFS_ROOT -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -device virtio-net-device,netdev=net0 -netdev user,id=net0,ipv6=off -append init_path=/sbin/startup-benchmark.sh' (10 runs):

       3787.537426      task-clock (msec)         #    0.858 CPUs utilized            ( +-  0.25% )
             1,134      context-switches          #    0.300 K/sec                    ( +-  0.43% )
                 2      cpu-migrations            #    0.001 K/sec                    ( +- 32.44% )
             7,632      page-faults               #    0.002 M/sec                    ( +-  0.32% )
    12,131,660,326      cycles                    #    3.203 GHz                      ( +-  0.90% )  (30.79%)
    34,744,544,393      instructions              #    2.86  insn per cycle           ( +-  0.21% )  (38.50%)
     4,178,763,151      branches                  # 1103.293 M/sec                    ( +-  0.15% )  (38.44%)
        48,974,386      branch-misses             #    1.17% of all branches          ( +-  0.26% )  (38.45%)
     8,994,262,144      L1-dcache-loads           # 2374.699 M/sec                    ( +-  0.15% )  (38.45%)
       116,658,814      L1-dcache-load-misses     #    1.30% of all L1-dcache hits    ( +-  1.05% )  (38.41%)
         3,721,859      LLC-loads                 #    0.983 M/sec                    ( +-  1.31% )  (30.68%)
           679,376      LLC-load-misses           #   18.25% of all LL-cache hits     ( +-  5.92% )  (30.67%)
   <not supported>      L1-icache-loads
       341,476,392      L1-icache-load-misses                                         ( +-  0.36% )  (30.72%)
     8,835,825,535      dTLB-loads                # 2332.868 M/sec                    ( +-  0.82% )  (30.80%)
           204,090      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +-  5.20% )  (30.89%)
         1,958,784      iTLB-loads                #    0.517 M/sec                    ( +-  1.71% )  (30.88%)
           705,677      iTLB-load-misses          #   36.03% of all iTLB cache hits   ( +-  8.49% )  (30.84%)

       4.414902419 seconds time elapsed                                          ( +-  0.16% )

jrtc27 commented on target/cheri-common/cheri_tagmem.c:
```diff
@@ -433,7 +564,14 @@ void cheri_tag_set(CPUArchState *env, target_ulong vaddr, int reg, hwaddr* ret_p
         RAM_ADDR_FMT "] %d -> 1\n", vaddr, ram_offset,
         tag_bit_get(ram_offset >> CAP_TAG_SHFT, ram));
-    tag_bit_set(ram_offset >> CAP_TAG_SHFT, ram);
+    if (tag_bit_set(ram_offset >> CAP_TAG_SHFT, ram)) {
```
Is the comment at the start of this function still true? It could certainly do with de-MIPS-ifying.

arichardson replied:
It does currently bypass the soft-TLB, but we should really change that. I'll update the comment once that is the case.

jrtc27 commented on target/cheri-common/cheri_tagmem.c:
```diff
@@ -433,7 +564,14 @@ void cheri_tag_set(CPUArchState *env, target_ulong vaddr, int reg, hwaddr* ret_p
         tag_bit_get(ram_offset >> CAP_TAG_SHFT, ram));
-    tag_bit_set(ram_offset >> CAP_TAG_SHFT, ram);
+    if (tag_bit_set(ram_offset >> CAP_TAG_SHFT, ram)) {
+        // new tag block allocated, flush the TCG tlb so that the magic zero
+        // value is removed from the iotlb.
+        // warn_report("Allocated new tag block for " TARGET_FMT_plx  "->
+        // flushing tlb\r", vaddr);
+        // TODO: assert that the value was the magic zero constant
```
Unless I'm missing something, this code is run whenever a tag is set, not just when we allocate a new tag block? So we'll be invalidating the TLB on every valid capability store?

arichardson replied:
It should only return true when a new block is allocated, so in the common case it won't flush.

jrtc27 replied:
Oh, I see. I feel that might be better as an optional out parameter...

arichardson replied:
Done in new version.

arichardson committed:
Use a magic constant for ALL_ZERO_TAGBLK instead of a global

This slightly improves code density for the check:
Instead of `mov $0x30e110,%ecx; cmp %rcx,%rax; je bd`, we now generate
`cmp $0x1,%rax;je bd`
It also means that all accidental writes to the magic "all tags are zero"
buffer will result in a crash rather than writing to memory that should
always be zero.

This also marginally improves performance due to one fewer instruction in the
hot path.
Before:
       3732.912310      task-clock (msec)         #    0.846 CPUs utilized            ( +-  0.24% )
             1,118      context-switches          #    0.300 K/sec                    ( +-  0.51% )
                 2      cpu-migrations            #    0.001 K/sec                    ( +- 14.16% )
             7,633      page-faults               #    0.002 M/sec                    ( +-  0.53% )
    12,415,263,274      cycles                    #    3.326 GHz                      ( +-  0.54% )  (30.76%)
    34,794,322,905      instructions              #    2.80  insn per cycle           ( +-  0.20% )  (38.53%)
     4,193,464,855      branches                  # 1123.376 M/sec                    ( +-  0.31% )  (38.56%)
        48,962,599      branch-misses             #    1.17% of all branches          ( +-  0.48% )  (38.60%)
     8,977,531,283      L1-dcache-loads           # 2404.967 M/sec                    ( +-  0.24% )  (38.64%)
       117,341,431      L1-dcache-load-misses     #    1.31% of all L1-dcache hits    ( +-  0.43% )  (38.55%)
         3,873,734      LLC-loads                 #    1.038 M/sec                    ( +-  1.97% )  (30.72%)
           659,584      LLC-load-misses           #   17.03% of all LL-cache hits     ( +-  9.49% )  (30.69%)
   <not supported>      L1-icache-loads
       339,522,757      L1-icache-load-misses                                         ( +-  0.53% )  (30.66%)
     8,906,369,663      dTLB-loads                # 2385.904 M/sec                    ( +-  0.33% )  (30.69%)
           236,073      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +-  3.57% )  (30.71%)
         2,058,548      iTLB-loads                #    0.551 M/sec                    ( +-  2.15% )  (30.71%)
           595,960      iTLB-load-misses          #   28.95% of all iTLB cache hits   ( +-  3.73% )  (30.70%)

       4.411608780 seconds time elapsed                                          ( +-  0.19% )

After:
       3743.966232      task-clock (msec)         #    0.851 CPUs utilized            ( +-  0.24% )
             1,128      context-switches          #    0.301 K/sec                    ( +-  0.30% )
                 3      cpu-migrations            #    0.001 K/sec                    ( +- 20.03% )
             7,603      page-faults               #    0.002 M/sec                    ( +-  0.68% )
    12,382,608,616      cycles                    #    3.307 GHz                      ( +-  0.43% )  (30.74%)
    34,665,991,103      instructions              #    2.80  insn per cycle           ( +-  0.30% )  (38.56%)
     4,173,745,114      branches                  # 1114.792 M/sec                    ( +-  0.24% )  (38.63%)
        48,537,899      branch-misses             #    1.16% of all branches          ( +-  0.37% )  (38.67%)
     8,939,423,205      L1-dcache-loads           # 2387.688 M/sec                    ( +-  0.20% )  (38.67%)
       118,483,071      L1-dcache-load-misses     #    1.33% of all L1-dcache hits    ( +-  0.56% )  (38.57%)
         3,923,479      LLC-loads                 #    1.048 M/sec                    ( +-  1.98% )  (30.66%)
           688,496      LLC-load-misses           #   17.55% of all LL-cache hits     ( +-  7.72% )  (30.63%)
   <not supported>      L1-icache-loads
       335,780,300      L1-icache-load-misses                                         ( +-  0.60% )  (30.58%)
     8,917,818,779      dTLB-loads                # 2381.918 M/sec                    ( +-  0.48% )  (30.69%)
           238,865      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +-  4.53% )  (30.70%)
         1,541,530      iTLB-loads                #    0.412 M/sec                    ( +-  3.22% )  (30.70%)
         1,063,661      iTLB-load-misses          #   69.00% of all iTLB cache hits   ( +-  5.93% )  (30.75%)

       4.400986730 seconds time elapsed                                          ( +-  0.22% )

arichardson committed:
Use an output parameter instead of a boolean return value

This makes the code slightly easier to read.

arichardson committed:
Correctly invalidate the TCG TLB when allocating a new tag block

The existing code had two bugs:
1) each tag block covers multiple pages so we need to invalidate all
   covered pages
2) Unless we drop out of the current translation block, the QEMU TLB will
   not be filled.

These bugs were not caught by booting CheriBSD, but fortunataly the MIPS
baremetal testsuite caught it and I added a new test there in
https://github.com/CTSRD-CHERI/cheritest/commit/ad9d932bf1a671740523d42941433f1fdf4c94db


Before (incorrect!):
       3743.966232      task-clock (msec)         #    0.851 CPUs utilized            ( +-  0.24% )
             1,128      context-switches          #    0.301 K/sec                    ( +-  0.30% )
                 3      cpu-migrations            #    0.001 K/sec                    ( +- 20.03% )
             7,603      page-faults               #    0.002 M/sec                    ( +-  0.68% )
    12,382,608,616      cycles                    #    3.307 GHz                      ( +-  0.43% )  (30.74%)
    34,665,991,103      instructions              #    2.80  insn per cycle           ( +-  0.30% )  (38.56%)
     4,173,745,114      branches                  # 1114.792 M/sec                    ( +-  0.24% )  (38.63%)
        48,537,899      branch-misses             #    1.16% of all branches          ( +-  0.37% )  (38.67%)
     8,939,423,205      L1-dcache-loads           # 2387.688 M/sec                    ( +-  0.20% )  (38.67%)
       118,483,071      L1-dcache-load-misses     #    1.33% of all L1-dcache hits    ( +-  0.56% )  (38.57%)
         3,923,479      LLC-loads                 #    1.048 M/sec                    ( +-  1.98% )  (30.66%)
           688,496      LLC-load-misses           #   17.55% of all LL-cache hits     ( +-  7.72% )  (30.63%)
   <not supported>      L1-icache-loads
       335,780,300      L1-icache-load-misses                                         ( +-  0.60% )  (30.58%)
     8,917,818,779      dTLB-loads                # 2381.918 M/sec                    ( +-  0.48% )  (30.69%)
           238,865      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +-  4.53% )  (30.70%)
         1,541,530      iTLB-loads                #    0.412 M/sec                    ( +-  3.22% )  (30.70%)
         1,063,661      iTLB-load-misses          #   69.00% of all iTLB cache hits   ( +-  5.93% )  (30.75%)

       4.400986730 seconds time elapsed                                          ( +-  0.22% )

After:
       3965.398176      task-clock (msec)         #    0.872 CPUs utilized            ( +-  0.41% )
             1,197      context-switches          #    0.302 K/sec                    ( +-  0.51% )
                 3      cpu-migrations            #    0.001 K/sec                    ( +- 20.83% )
             9,886      page-faults               #    0.002 M/sec                    ( +- 22.31% )
    12,975,511,287      cycles                    #    3.272 GHz                      ( +-  0.33% )  (30.96%)
    36,310,805,061      instructions              #    2.80  insn per cycle           ( +-  0.32% )  (38.68%)
     4,290,656,596      branches                  # 1082.024 M/sec                    ( +-  0.31% )  (38.60%)
        52,444,394      branch-misses             #    1.22% of all branches          ( +-  0.41% )  (38.54%)
     9,520,541,071      L1-dcache-loads           # 2400.904 M/sec                    ( +-  0.27% )  (38.44%)
       121,941,572      L1-dcache-load-misses     #    1.28% of all L1-dcache hits    ( +-  0.84% )  (38.36%)
         3,899,538      LLC-loads                 #    0.983 M/sec                    ( +-  6.62% )  (30.56%)
           904,366      LLC-load-misses           #   23.19% of all LL-cache hits     ( +- 21.52% )  (30.59%)
   <not supported>      L1-icache-loads
       338,198,675      L1-icache-load-misses                                         ( +-  0.74% )  (30.66%)
     9,544,742,562      dTLB-loads                # 2407.007 M/sec                    ( +-  0.30% )  (30.69%)
           220,078      dTLB-load-misses          #    0.00% of all dTLB cache hits   ( +- 11.12% )  (30.84%)
         4,496,137      iTLB-loads                #    1.134 M/sec                    ( +-  1.62% )  (30.87%)
           681,771      iTLB-load-misses          #   15.16% of all iTLB cache hits   ( +- 15.80% )  (30.90%)

       4.547806136 seconds time elapsed                                          ( +-  0.24% )

Even with this change it is still significantly faster than the baseline.

arichardson commented:
I also tried calling mmap() to allocate all of tag memory, but that appears to be significantly slower than the current approach (even with all the TCG tlb flushing).

jrtc27 commented on target/cheri-common/cheri_tagmem.c:
```diff
@@ -313,67 +318,197 @@ static inline RAMBlock *v2r_addr(CPUArchState *env, target_ulong vaddr,
+    abort();
+#endif
+    assert(p_old == p_new);
+#if 0
```
This probably doesn't want committing?

arichardson replied:
Will update tomorrow

jrtc27 reviewed (commented):
Is there a reason why this has stalled beyond nobody commenting on it?

arichardson commented:
> Is there a reason why this has stalled beyond nobody commenting in it?

Yeah I was waiting for review but it wasn't high priority so I forgot about this PR

jrtc27 commented on target/cheri-common/cheri_tagmem.c:
```diff
@@ -313,67 +318,178 @@ static inline RAMBlock *v2r_addr(CPUArchState *env, target_ulong vaddr,
+{
+    return offset.value << CAP_TAG_SHFT;
+}
+
+
```
```suggestion

```

jrtc27 reviewed (commented):
Looks plausible with my limited QEMU knowledge, just a whitespace nit

nwf commented:
LGTM; sorry for not reviewing much earlier
